pinot-transport/src/main/java/com/linkedin/pinot/transport/netty/NettyServer.java

```diff
@@ -193,69 +191,74 @@ public NettyChannelInboundHandler(RequestHandler handler, NettyServerMetrics met
       this(handler, metric, 100);
     }
 
-    public enum State {
-      INIT,
-      REQUEST_RECEIVED,
-      RESPONSE_WRITTEN,
-      RESPONSE_SENT,
-      EXCEPTION
-    }
-
-    /**
-     * Server Channel Handler State
-     */
-    private State _state = State.INIT;
-
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) {
-      _requestStartTime = System.currentTimeMillis();
+      final long requestStartTime = System.currentTimeMillis();
       LOGGER.debug("Request received by server !!");
-      _state = State.REQUEST_RECEIVED;
-      ByteBuf request = (ByteBuf) msg;
-      _lastRequsetSizeInBytes = request.readableBytes();
+
+      final ByteBuf request = (ByteBuf) msg;
+      final long requestSizeInBytes = request.readableBytes();
 
       //Call processing handler
-      _lastProcessingLatency = MetricsHelper.startTimer();
-      byte[] response = _handler.processRequest(ctx, request);
-      _lastProcessingLatency.stop();
-
-      // Send Response
-      ByteBuf responseBuf = Unpooled.wrappedBuffer(response);
-      _lastSendResponseLatency = MetricsHelper.startTimer();
-      ChannelFuture f = ctx.writeAndFlush(responseBuf);
-      _state = State.RESPONSE_WRITTEN;
-      f.addListener(this);
-      request.release();
+      final TimerContext requestProcessingLatency = MetricsHelper.startTimer();
+      final ChannelHandlerContext requestChannelHandlerContext = ctx;
+      ListenableFuture<byte[]> serializedQueryResponse = _handler.processRequest(ctx, request);
+      Futures.addCallback(serializedQueryResponse, new FutureCallback<byte[]>() {
```

Comment:
when will the FutureCallback get invoked? is it internally put into another  queue and can be executed in any order. 
My main concern is we have already processed the request and response is ready but instead of sending it right away we put it in another executor service to send the response back. 

I haven't looked at the internals of the implementation but java docs say "This overload, which does not accept an executor, uses directExecutor, a dangerous choice in some cases. See the discussion in the ListenableFuture.addListener documentation.".
Looks like directExecutor means execute runnable in the same thread. Why do they say it can be dangerous?

Also, this code can benefit with some formatting. It's very hard to parse especially with the anonymous inner classes in the arguments.